### DISCUSSION TITLE
Please walk me through the implementation steps

### DISCUSSION THREAD

**Discussion by sansi-zhang** [Xilinx/finn#1049] [Q&A]
Hi, @maltanar

Now I hope not to use the ''Finn'' framework, but to use the hardware optimization in ''Finn'' to implement some network structures through hls that are not currently supported in the ''Finn'' architecture.

My idea was to write 'hls' for my network through the code in 'finn-hlslib', but I didn't fully understand the layout and meaning of the code, and the documentation was no longer valid.

Therefore, I would like to ask for your help: can you recommend me some information about 'finn-hlslib', and can you help me sort out my thinking in development (I use the ZCU104 development board)?

Thank you so much.

**Comment by preusser** [comment]
Hi @sansi-zhang,
We are currently deemphasizing the exposure of the FINN HLSlib to FINN users. For the construction of efficient hardware, we increasingly rely on thoroughly engineered RTL solutions for bread-and-butter operators. We also want to claim the freedom for a serious refactoring of the HLSlib itself. It has accumulated quite some difficult ballast originating deep in the history of FINN, which started out as an inference solution for fully binarized neural network and has been extended from there without the mechanisms now available in C++14 and the latest Vitis versions.
Nonetheless, I'd be happy to point you to the HLS implementations of individual operators that you might need if you can specified what functionality you are interested in.

**Reply by sansi-zhang** [reply]
Thank you very much.

Now I want to realize the design of MVU, VVU and SWU in 'finn' through hls to realize the deep learning network design in the image field.

Since I now only understand the simple implementation of deep learning network through hls, I can not clearly understand how to achieve these optimization designs.

I tried to read the relevant implementation code in finn-hlslib, but I didn't get a good understanding.

If possible, I hope you can give me some guidance, thanks again.

**Reply by sansi-zhang** [reply]
Hi @preusser,
I'm sorry to bother you again.
You haven't replied to my last question.
If you have time, I hope you can give me an answer.

Specifically, I need to know how to manually implement the design of MVU, PE and VVU structures through hls.

Thanks.
